src/manifolds/StiefelCanonicalMetric.jl

```diff
@@ -0,0 +1,115 @@
+We introduce ``A=p^{\mathrm{T}}X`` and ``QR = (I_n-pp^{\mathrm{T}})X``` the qr decomposition
+of the vertical component. Then using the matrix exponential ``operatorname{Exp}`` we introduce ``B`` and ``C`` as
+
+```math
+\begin{pmatrix}
+B\\C
+\end{pmatrix}
+\coloneqq
+\operatorname{Exp}\left(
+\begin{pmatrix}
+A & -R^{\mathrm{T}}\\ R & 0
+\end{pmatrix}
+\right)
+\begin{pmatrix}I_k\\0\end{pmatrix}
+```
+
+the exponential map reads
+
+```math
+q = \exp_p X = pC + QB.
+```
+For more details, see [^EdelmanAriasSmith1998][^Zimmermann2017].
+"""
+exp(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, ::Any...) where {n,k}
+
+function exp!(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, Q, p, X) where {n,k}
+    A = p' * X
+    QR = qr(X - p * A)
+    BC_ext = exp([A -QR.R'; QR.R, zeros(k, k)])
+    q .= [p Matrix(QR.Q)] * BC_ext[:, 1:k]
+    return q
+end
+
+@doc raw"""
+    inner(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, X, CanonicalMetric}, p,, X, Y)
+
+compute the inner procuct on the [`Stiefel`](@ref) manifold with respect to the
+[`CanonicalMetric`](@ref). The formula reads
+
+```
+g_p(X,Y) = \operatorname{tr}\bigl( X^{\mathrm{T}}(I_n - \frac{1}{2}pp^{\mathrm{T}})Y \bigr).
+```
+"""
+function inner(::MetricManifold{ℝ,Stiefel{n,k,ℝ},X,CanonicalMetric}, p, X, Y) where {n,k}
+    return tr(X' * (Matrix(I, n, n) - 0.5 * p * p') * Y)
+end
+
+@doc raw"""
+    X = log(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, CanonicalMetric}, p, q)
+    log!(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, X, CanonicalMetric}, p, q)
+
+Compute the logarithmic map on the [`Stiefel`](@ref)`(n,k)` manifold with respect to the [`CanonicalMetric`](@ref)
+using a matrix-algebraic based approach to an iterative inversion of the formula of the
+[`exp`](@ref exp(::MetricManifold{ℝ, Stiefel{n,k,ℝ}, CanonicalMetric}, ::Any...)).
+
+The algorithm is derived in[^Zimmermann2017].
+
+[^Zimmermann2017]:
+    > Zimmermann, R.: _A matrix-algebraic algorithm for the Riemannian logarithm on the Stiefel manifold under the canoncial metric.
+    > SIAM Journal on Matrix Analysis and Applications 28(2), pp. 322-342, 2017.
+    > doi: [10.1137/16M1074485](https://doi.org/10.1137/16M1074485),
+    > arXiv: [1604.05054](https://arxiv.org/abs/1604.05054).
+"""
+log(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, ::Any...) where {n,k}
+
+function log!(
+    ::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric},
+    p,
+    q;
+    tolerance=1e-9,
+    maxiter=1e5,
+) where {n,k}
+    M = p' * q
+    QR = qr(q - p * M)
+    V = qr([M; Matrix(QR.R)]).Q * Matrix(I, 2 * k, 2 * k)
+    S = svd(V[(k + 1):(2 * k), (k + 1):(2k)]) #bottom right corner
+    V[:, (k + 1):(2 * k)] = V[:, (k + 1):(2 * k)] * (S.V * S.U')
+    LV = log(V)
+    C = view(LV, (k + 1):(2 * k), (k + 1):(2 * k))
+    i = 0
+    while (i < maxiter) && (norm(C) > tolerance)
+        LV = log(V)
+        V[:, (k + 1):(2 * k)] *= exp(-C)
+    end
+    X .= p * LV[1:k, 1:k] + QR.Q * LV[(k + 1):(2 * k), 1:k]
```

Comment:
Again we can eliminate allocations here
```suggestion
    mul!(X, p, @view(LV[1:k, 1:k]))
    mul!(X, QR.Q, @view(LV[(k + 1):(2 * k), 1:k]), true, true)
```